### ISSUE TITLE
Notebook fails to reconnect on flaky network

### ISSUE THREAD

**Issue by aggFTW** [jupyter/notebook#1379]
On flaky network connections, the notebook will show the "Connection failed" dialog.

The Console shows that websocket connections fail with “Error in connection establishment: net::ERR_NAME_NOT_RESOLVED”.

Even though the Console might say that "kernel_connected" again, the status bar will remain disconnected (broken chain icon next to kernel name) and the cells will show as executing (busy start next to cell) if you try to execute them but the kernel won't receive any messages.

**Comment by aggFTW** [comment]
cc @ellisonbg

**Comment by ellisonbg** [comment]
I am visiting @aggFTW today - he said that this is reproducible. He will describe steps needed to reproduce. Probably one for @willingc or @minrk

**Comment by ellisonbg** [comment]
I tried to reproduce on master, but couldn't. They will provide more details.

**Comment by aggFTW** [comment]
I was able to repro with 4.1 and 4.2. Steps:

1. Connect from your browser to a remote notebook instance, or to an instance that you can temporarily halt network connectivity to.
2. Create a notebook.
3. Remove network connectivity so that your browser cannot talk to the kernel.
4. Try to execute a cell without network connectivity.
5. Wait for the `Connection failed` message to appear in the UI. By then, the console should show WebSocket connection to kernel failed and that it's retrying.
6. Reestablish network connectivity.
7. Console should say that kernel_connected.
8. Click OK in `Connection failed` dialogue.
9. Observe that disconnected icon is still in the status bar and cell is still supposedly executing.
10. New cell cannot be executed (maybe the frontend thinks the first cell hasn't finished execution and there's no way for the frontend to reconcile the state?)
11. Refresh your browser as the only way to get out of this state.

**Comment by willingc** [comment]
@aggFTW Thanks for the detailed steps to reproduce. I didn't get exactly the same results as you did but what I did get was odd. 

@minrk There's some sort of issue here. I ran a notebook server 4.1 on an Ubuntu Digital Ocean instance. I connected from my laptop to the server, created a notebook, and then entered and ran 2 cells. I entered a third cell and before executing I turned off my laptop wifi to simulate a network outage. 

I waited quite a while for the client to time out and display Connection failed; yet, it never appeared. What I did finally get was an error icon after I reconnected the wifi. When I executed the third cell this was the output with the third cell being numbered 1:

<img width="987" alt="screen shot 2016-05-03 at 9 41 12 pm" src="https://cloud.githubusercontent.com/assets/2680980/15005451/a299f1fa-1179-11e6-9287-6a3d18fbdc0d.png">

Here's the [pastebin](http://pastebin.com/KVpnXB0f) of the debug console.

I'll play around with it a bit more tomorrow. @aggFTW, would you mind running the process you followed again with `--debug` when starting the notebook server and posting the output? Thanks.

**Comment by ellisonbg** [comment]
@willingc thanks for looking at this! I am workiing with @aggFTW today at
MSFT - let us know if there is anything we can do to help debug this.

On Tue, May 3, 2016 at 10:02 PM, Carol Willing <notifications@github.com>
wrote:

> @aggFTW <https://github.com/aggFTW> Thanks for the detailed steps to
> reproduce. I didn't get exactly the same results as you did but what I did
> get was odd.
>
> @minrk <https://github.com/minrk> There's some sort of issue here. I ran
> a notebook server 4.1 on an Ubuntu Digital Ocean instance. I connected from
> my laptop to the server, created a notebook, and then entered and ran 2
> cells. I entered a third cell and before executing I turned off my laptop
> wifi to simulate a network outage.
>
> I waited quite a while for the client to time out and display Connection
> failed; yet, it never appeared. What I did finally get was an error icon
> after I reconnected the wifi. When I executed the third cell this was the
> output with the third cell being numbered 1:
>
> [image: screen shot 2016-05-03 at 9 41 12 pm]
> <https://cloud.githubusercontent.com/assets/2680980/15005451/a299f1fa-1179-11e6-9287-6a3d18fbdc0d.png>
>
> Here's the pastebin <http://pastebin.com/KVpnXB0f> of the debug console.
>
> I'll play around with it a bit more tomorrow. @aggFTW
> <https://github.com/aggFTW>, would you mind running the process you
> followed again with --debug when starting the notebook server and posting
> the output? Thanks.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/jupyter/notebook/issues/1379#issuecomment-216743450>
>



-- 
Brian E. Granger
Associate Professor of Physics and Data Science
Cal Poly State University, San Luis Obispo
@ellisonbg on Twitter and GitHub
bgranger@calpoly.edu and ellisonbg@gmail.com

**Comment by aggFTW** [comment]
Thanks for trying that @willingc!

About
> I entered a third cell and before executing I turned off my laptop wifi to simulate a network outage.
> I waited quite a while for the client to time out and display Connection failed; yet, it never appeared

Did you execute the cell after you simulated the network outage? After executing the cell, it doesn't take too long for the message to appear.

**Comment by willingc** [comment]
@aggFTW In an incognito window in Chrome I got essentially your described response. While in a regular window, I got similar response but also got a response that worked as it was supposed to work.

I'll take a look at the logs tomorrow. Feel free to send along any logs that you  may have. Thanks!

**Comment by aggFTW** [comment]
So I ran this with `--debug`, and looking at my Network/Console in my browser.

Timeline:
* `[D 2016-05-07 01:55:43.855 NotebookApp] 200 GET /notebooks/PySpark/Untitled.ipynb (10.0.0.26) 62.31ms`
* `[D 2016-05-07 01:55:44.922 NotebookApp] Starting kernel: [u'python', u'-m', u'remotespark.kernels.pysparkkernel.pysparkkernel'...`
* `[D 2016-05-07 01:55:45.021 NotebookApp] Initializing websocket connection /api/kernels/bf2dcf74-056e-439b-84d1-0de1e6bd95ad/channels`
* Initial websocket handshake at `Sat, 07 May 2016 01:55:46 GMT` from browser headers
* From server, see logs:
```
[D 2016-05-07 01:55:47.288 NotebookApp] Opening websocket /api/kernels/bf2dcf74-056e-439b-84d1-0de1e6bd95ad/channels
[D 2016-05-07 01:55:47.289 NotebookApp] Connecting to: tcp://127.0.0.1:46299
[D 2016-05-07 01:55:47.289 NotebookApp] Connecting to: tcp://127.0.0.1:52057
[D 2016-05-07 01:55:47.290 NotebookApp] Connecting to: tcp://127.0.0.1:33979
```
* Manually disabled network and executed cell. Waited for `Error in connection establishment` errors from Console from `WebSocket connection`
* Enabled network again
* Reestablished websocket handshake at `Sat, 07 May 2016 01:56:50 GMT` from browser headers
* From server, see logs:
```
[D 2016-05-07 01:56:51.545 NotebookApp] Initializing websocket connection /api/kernels/bf2dcf74-056e-439b-84d1-0de1e6bd95ad/channels
[D 2016-05-07 01:56:51.550 NotebookApp] Opening websocket /api/kernels/bf2dcf74-056e-439b-84d1-0de1e6bd95ad/channels
[D 2016-05-07 01:56:51.550 NotebookApp] Connecting to: tcp://127.0.0.1:46299
[D 2016-05-07 01:56:51.551 NotebookApp] Connecting to: tcp://127.0.0.1:52057
[D 2016-05-07 01:56:51.551 NotebookApp] Connecting to: tcp://127.0.0.1:33979
```

After that, no activity on websockets from the browser (checked network and console). Notebook remains with broken chain icon:
![a](https://cloud.githubusercontent.com/assets/654204/15089766/4ff2e22e-13c0-11e6-97fd-bb37e481c07d.png)
and notebook cell appears to be busy and new cells are just queued but not executed:
![b](https://cloud.githubusercontent.com/assets/654204/15089774/771bbcfe-13c0-11e6-8b27-ed034e8f6d1e.png)

Upon browser refresh, notebook execution is normal.

I hope that's helpful :smile:

**Comment by aggFTW** [comment]
Hi @willingc and @ellisonbg. Any updates on this? How else can we help?
Thanks!

**Comment by willingc** [comment]
@minrk Do you have any ideas why a network outage would give this behavior? I can't tell if it's a timing issue or other. 

cc @aggFTW

**Comment by minrk** [comment]
@aggFTW does it work if you select reconnect from the Kernel menu?

<img width="194" alt="screen shot 2016-05-11 at 11 07 11" src="https://cloud.githubusercontent.com/assets/151929/15175864/9fc9b3ee-1768-11e6-8b6b-daa5a198c6a2.png">

**Comment by aggFTW** [comment]
@minrk No, Kernel -> Reconnect leaves the notebook in the same non-functional state.

**Comment by aggFTW** [comment]
@ellisonbg @minrk @willingc Any updates? :smiley:

**Comment by minrk** [comment]
@aggFTW I believe this should be fixed by #1471.

**Comment by willingc** [comment]
Thanks @minrk and @aggFTW for resolving the reconnect.
